File: JAVA/AccessModifiers/src/com/abhisheksingh/Account.java
package com.abhisheksingh;

import java.util.ArrayList;

public class Account {
    private String accountName;
    private double balance;
    private ArrayList<Double> transactions;

    {
        System.out.println("Initializer Block 2 called");
    }

    {
        System.out.println("Initializer Block called");
    }


    public Account(String accountName) {
        this.accountName = accountName;
        this.transactions = new ArrayList<>();
        this.balance = 0.0;
    }

    public double getBalance() {
        return balance;
    }

    public void deposit(double amount) {
        if(amount > 0.0) {
            transactions.add(amount);
            balance += amount;
            System.out.println(amount + " deposited successfully. The balance is now : " + balance);
        }
        else {
            System.out.println("Cannot deposit negative amount.");
        }
    }

    public void withdraw(double amount) {
        if(amount < 0.0 && (balance - amount < 0.0)) {
            System.out.println("Cannot withdraw the amount.");
        }
        else {
            transactions.add(-amount);
            balance -= amount;
            System.out.println(amount + " withdraw successful. The balance is now : " + balance);
        }
    }
}

File: cpp_course/shared_pointers/src/Account.cpp
#include "Account.h"

Account::Account(std::string name, double balance)
    : name {name}, balance {balance}
{}

bool Account::deposit(double amount)
{
    if(amount < 0)
        return false;
    
    balance += amount;
    return true;
}

bool Account::withdraw(double amount)
{
    if((balance - amount >= 0) && (amount > 0))
    {
        balance -= amount;
        return true;
    }

    return false;
}


bool Account::operator+=(double amount)
{
    return deposit(amount);
}  

bool Account::operator-=(double amount)
{
    return withdraw(amount);
}   